### PR TITLE
Byte string parser input for request target parsing

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

**Comment by johanandren** [comment]
Local JMH scores:

**baseline**
```
[info] Benchmark                                                 Mode  Cnt       Score       Error  Units
[info] HttpRequestParsingBenchmark.parse_10000_single_requests  thrpt   20  553248,512 ± 13052,126  ops/s
[info] HttpRequestParsingBenchmark.parse_1000_single_requests   thrpt   20  533321,279 ±  2603,929  ops/s
[info] HttpRequestParsingBenchmark.parse_100_single_requests    thrpt   20  437281,331 ±  2794,637  ops/s
```
**ByteStringParserInput** 
```
[info] Benchmark                                                 Mode  Cnt       Score      Error  Units
[info] HttpRequestParsingBenchmark.parse_10000_single_requests  thrpt   20  554082,668 ± 7487,084  ops/s
[info] HttpRequestParsingBenchmark.parse_1000_single_requests   thrpt   20  528730,408 ± 2138,058  ops/s
[info] HttpRequestParsingBenchmark.parse_100_single_requests    thrpt   20  442216,722 ± 4157,171  ops/s
```

So no real perf improvement, possibly a little bit lower gc overhead. Is a tad cleaner than the original version with `asArray` though.

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM.

Warning, crazy idea incoming:


```scala
trait ParserInput extends Any {
 ...
```

```scala
final class ByteStringParserInput(val bytes: ByteString) extends AnyVal with ParserInput {
  override def charAt(ix: Int): Char = (bytes(ix) & 0xFF).toChar
  override def length: Int = bytes.size
  override def sliceString(start: Int, end: Int): String = bytes.slice(start, end - start).decodeString(StandardCharsets.ISO_8859_1)
  override def sliceCharArray(start: Int, end: Int): Array[Char] =
    StandardCharsets.ISO_8859_1.decode(bytes.slice(start, end).asByteBuffer).array()
  def getLine(line: Int): String = {
    @tailrec def rec(ix: Int, lineStartIx: Int, lineNr: Int): String =
      if (ix < length)
        if (charAt(ix) == '\n')
          if (lineNr < line) rec(ix + 1, ix + 1, lineNr + 1)
          else sliceString(lineStartIx, ix)
        else rec(ix + 1, lineStartIx, lineNr)
      else if (lineNr == line) sliceString(lineStartIx, ix) else ""
    rec(ix = 0, lineStartIx = 0, lineNr = 1)
  }

}
```

zero allocations, yay

File: akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package akka.http.impl.util
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala]: copyright ;)

Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala]: darn

**Comment by johanandren** [comment]
Tried the optimized `getLine` no visible improvement.

**Comment by ktoso** [comment]
> Tried the optimized getLine no visible improvement.

You mean the ValueClass hint? The in-lined getLine is the same, but we're forced to copy it as the other trait is not an universal trait.

**Comment by johanandren** [comment]
Ah, missed the `AnyVal` part! Will try again.

**Comment by johanandren** [comment]
Nope, not possible. `DefaultParserInput` is an abstract class

**Comment by johanandren** [comment]
haha, loop: Ah, missed the `ParserInput` part! Will try again.

**Comment by johanandren** [comment]
No cigar anyways, `ParserInput` would have to have `AnyVal` as superclass.

**Comment by ktoso** [comment]
Which we can try ;) And it should be Any, not AnyVal, an universal trait

**Comment by ktoso** [comment]
Checked the other crazy ideas btw, they don't change a thing.
LGTM

**Comment by ktoso** [comment]
Resolves https://github.com/akka/akka/issues/21039

**Pull request merged**